actionpack/lib/action_dispatch/middleware/executor.rb

```diff
@@ -13,6 +13,9 @@ def call(env)
       begin
         response = @app.call(env)
         returned = response << ::Rack::BodyProxy.new(response.pop) { state.complete! }
+      rescue => error
+        @executor.error_reporter.report(error, handled: false)
```

Comment:
I'll have a look. But the idea what that you'd either register into `Rails.error` OR insert the middleware, not both.

> So it'd be great to have something like Rails.error_reporter.disable(my_subscriber) {

Can't you implement `my_subscriber.disable { }`?

Reply:
> do it in Sidekiq's ErrorHandler for the contextual data.

If Sidekiq gives you some context `Rails.error`, then yes the `disable` may make sense. Could you expand on what info it gives?

> Another thing is that, having a subscriber registered doesn't mean all the exceptions will be captured by it, or is it? From what I can tell, the covered areas are:

Well, all exceptions from anything wrapped in the executor, which is supposed to be everything. We might miss a few things like `rails runner`, but `ActiveJob` and `Rack` is covered by it, which should cover 99% of use cases .

Do you have some specific issues in mind causing some code not to be covered?

> we don't always know if the execution will be wrapped with executor.wrap or not.

The idea is that any user code should be wrapped. If it isn't it mean some uncommon entry point was used and we should try to cover it.

> But I raised it here because I think may not be a sentry-specific problem and having a Rails-level support will help us maintainers a lot

If it really makes your live easier, then sure, no problem, I can add it. Won't be there before 7.1 though.

Reply:
> Could you expand on what info it gives?

Because `executor.wrap` is ran inside Sidekiq's middlewares, which we use to [add most of the contextual data to Sentry's scope](https://github.com/getsentry/sentry-ruby/blob/master/sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb#L16-L21) (similar to Rails' `ExecutionContext`), the `Rails.error` report actually already has a bunch of info:

<img width="70%" alt="截圖 2022-02-18 16 04 27" src="https://user-images.githubusercontent.com/5079556/154718508-7a4a83a6-70e4-42b6-8445-8c5e98b625df.png">

And this is what `sentry-sidekiq` captures:

<img width="80%" alt="截圖 2022-02-18 16 05 37" src="https://user-images.githubusercontent.com/5079556/154718633-9cd95fe9-b3f7-463f-b494-33ac179f6708.png">

So the differences are:
- `context` - which indicates if the error is caused by job execution or internal exception ([source](https://github.com/mperham/sidekiq/blob/46f4c6bd9ee0b2f5502a1dc0d03414989ad65063/lib/sidekiq/processor.rb#L177)). I think this is neglect-able in most cases.
- `job_str` - the raw string of the job, which in most cases won't be needed either.

(Looks like we're doing a goob job providing contextual info to the scope 🤔)

But I think this is not the case for all SDKs. For example, it looks like [appsignal-ruby](https://github.com/appsignal/appsignal-ruby/blob/main/lib/appsignal/integrations/sidekiq.rb#L11-L32) doesn't pre-store contextual data for every job. And in their case the `Rails.error` report should have a lot less data than the one captured by their SDK.

Of course this should be reported by them, not by me. Sentry is probably fine to let `Rails.error` do the report.

> Well, all exceptions from anything wrapped in the executor, which is supposed to be everything. We might miss a few things like rails runner, but ActiveJob and Rack is covered by it, which should cover 99% of use cases .

> The idea is that any user code should be wrapped. If it isn't it mean some uncommon entry point was used and we should try to cover it.

I think `resque` doesn't have `executor.wrap` in their codebase (at least I can't find with a code search). Is this expected?

> If it really makes your live easier, then sure, no problem, I can add it. Won't be there before 7.1 though.

I personally think it's good to let SDKs decide when to opt-out their own subscriber. So I'm looking forward to have it. But we can also wait for other maintainers feedback on this. I think most of them haven't adopted this feature yet.

In the mean time I'll also add some de-dup mechanism to the SDK. Thanks for your quick response 👍

Reply:
> doesn't pre-store contextual data for every job

Yeah, but it's also the job of Active Job adapters to forward whatever might be relevant into `ActiveSupport::ExecutionContext` so that it's available for the error reporter. That's something for us to improve.

My goal ultimately is that you should have all this info without having to hook into sidekiq specific API.

> I think resque doesn't have executor.wrap in their codebase (at least I can't find with a code search). Is this expected?

Active Job does it, no need for the backend to do it. I'm pretty sure Sidekiq could stop calling the reloader, I'd have to look at it.

> Thanks for your quick response 👍

Welcome. Thank you for your feedback, I very much intent to continue improve this API. If it's not good enough yet, we'll get another shot in 7.1 :)

Reply:
Feel free to ping me for ideas or feedback 😄

Reply:
@casperisfine I just found another issue related to this rescue block:

Sidekiq uses [`Sidekiq::JobRetry::Skip`](https://github.com/mperham/sidekiq/blob/e7d154eeb5e237a6e6e591a3bfca0a8f7bce32f6/lib/sidekiq/job_retry.rb#L65-L67) for its internal flow control. It [acts as a container to carry the actual exception](https://github.com/mperham/sidekiq/blob/e7d154eeb5e237a6e6e591a3bfca0a8f7bce32f6/lib/sidekiq/processor.rb#L172-L178) and doesn't mean to be reported. 

But it's now captured by this rescue block:

<img width="60%" alt="截圖 2022-03-06 12 39 07" src="https://user-images.githubusercontent.com/5079556/156923675-d5560d5e-e880-4247-a81d-029148fb731b.png">

I'm not sure how this could be solved. Perhaps Sidekiq can stop wrapping the execution for Rails 7+ apps?

cc @mperham